pkg/daemon/certificate_writer.go

```diff
@@ -101,92 +114,317 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
+			if cmErr != nil {
+				klog.Errorf("Error retrieving kubeconfig-data. %v", cmErr)
+			} else {
+				data, err = cmToData(cm, "ca-bundle.crt")
 				if err != nil {
-					mode = defaultDirectoryPermissions
-				} else {
-					mode = Dinfo.Mode()
+					klog.Errorf("kubeconfig-data ConfigMap not populated yet. %v", err)
+				} else if data != nil {
+					currentKC := clientcmdv1.Config{}
+					kcBytes, err := os.ReadFile(kubeConfigPath)
+					if err != nil {
+						return err
+					}
+					if kcBytes != nil {
+						err = yaml.Unmarshal(kcBytes, &currentKC)
+						if err != nil {
+							return fmt.Errorf("could not unmarshal kubeconfig into struct. Data: %s, Error: %v", string(kcBytes), err)
+						}
+						kubeConfigDiff = !bytes.Equal(bytes.TrimSpace(currentKC.Clusters[0].Cluster.CertificateAuthorityData), bytes.TrimSpace(data))
+						logSystem("Current data %s and New Data %s", string(currentKC.Clusters[0].Cluster.CertificateAuthorityData), string(data))
+						// if bytes.Compare is 1 AND theres no new data we do not want to write to disk. Just keep the old contents they will work.
+						// disk (currentKC) should be longer than configmap (data)
+						// we need to get the current data (kubeconfig-data) and compare it to clusters[0] on disk
+						// if ALL of kubeconfig-data (per cert) is contained on disk, do nothing
+						numericalDiff := bytes.Compare(currentKC.Clusters[0].Cluster.CertificateAuthorityData, data)
+						// sometimes it seems some of the data here might be invalid.
+						if numericalDiff == 1 && len(currentKC.Clusters[0].Cluster.CertificateAuthorityData) > 0 {
+							logSystem("Detecting deleted data from kubeconfig. Checking if we need to write to disk.")
+							certsConfigmap := strings.SplitAfter(strings.TrimSpace(string(data)), "-----END CERTIFICATE-----")
+							certsDisk := strings.SplitAfter(strings.TrimSpace(string(currentKC.Clusters[0].Cluster.CertificateAuthorityData)), "-----END CERTIFICATE-----")
+							// worst case scenario we are doubling the length
+							FullCA = make([]string, len(certsDisk)+len(certsConfigmap))
+							// write the whole old KC to this array. Then splice in new certs.
+							for i, ca := range certsDisk {
+								FullCA[i] = ca
+							}
+							for i, cert := range certsConfigmap {
+								// for each cert in the CM, we either ignore it OR splice it into the CertificateAuthorityData
+								// if on disk contains this cert, we are good
+								// if not, then that means we are adding certs.
+								if strings.Contains(strings.TrimSpace(string(currentKC.Clusters[0].Cluster.CertificateAuthorityData)), cert) {
+									continue
+								}
+								// we need to check for the subject.
+								b, _ := pem.Decode([]byte(cert))
+								if b == nil {
+									klog.Infof("Unable to decode cert %s into a pem block. Cert is either empty or invalid.", cert)
```

Comment:
looking at the logs, this or the next message is triggering for the `openshift-kube-apiserver-operator_localhost-recovery-serving-signer@1709645028`, which looks like a valid cert based on what you decoded. I wonder if that's normal?

Reply:
I think its fine, we still write it to disk. Will be good to look into though